File: leetcode3/염혜정/1665. Minimum Initial Energy to Finish Tasks.java
// o(n log n)
// 최소 필요 에너지 - 실제 소비 에너지가 큰 작업부터 처리

import java.util.Arrays;

class Solution {
    public int minimumEffort(int[][] tasks) {
        Arrays.sort(tasks, (a, b) -> {
            int gapA = a[1] - a[0];
            int gapB = b[1] - b[0];
            return Integer.compare(gapB, gapA);
        });

        int minimumEnergy = 0;
        int spentEnergy = 0;

        for (int[] task : tasks) {
            int actual = task[0];
            int minimum = task[1];
            minimumEnergy = Math.max(minimumEnergy, spentEnergy + minimum);
            spentEnergy += actual;
        }

        return minimumEnergy;
    }
}

File: leetcode3/최원준/1665. Minimum Initial Energy to Finish Tasks.py
class Solution:
    def minimumEffort(self, tasks: list[list[int]]) -> int:
        tasks.sort(key=lambda x: x[1] - x[0], reverse=True)

        ans = 0
        curr = 0

        for actual, minimum in tasks:
            ans = max(ans, curr + minimum)
            curr += actual

        return ans
